libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,765 @@
+    case ',':
+        // end of a term, add to checksum
+        _checksum ^= c;
+        FALLTHROUGH;
+    case '\r':
+    case '\n':
+    case '*':
+    {
+        if (_sentence_done) {
+            return false;
+        }
+
+        // null terminate and decode latest term
+        _term[_term_offset] = 0;
+        bool valid_sentence = decode_latest_term();
+
+        // move onto next term
+        _term_number++;
+        _term_offset = 0;
+        _term_is_checksum = (c == '*');
+        return valid_sentence;
+    }
+
+    case '!': // sentence begin
+        _sentence_valid = false;
+        _term_number = 0;
+        _term_offset = 0;
+        _checksum = 0;
+        _term_is_checksum = false;
+        _sentence_done = false;
+        return false;
+    }
+
+    // ordinary characters are added to term
+    if (_term_offset < sizeof(_term) - 1) {
+        _term[_term_offset++] = c;
+    }
+    if (!_term_is_checksum) {
+        _checksum ^= c;
+    }
+
+    return false;
+}
+
+// decode the most recently consumed term
+// returns true if new sentence has just passed checksum test and is validated
+bool AP_AIS::decode_latest_term()
+{
+    // handle the last term in a message
+    if (_term_is_checksum) {
+        _sentence_done = true;
+        uint8_t checksum = 16 * char_to_hex(_term[0]) + char_to_hex(_term[1]);
+        return ((checksum == _checksum) && _sentence_valid);
+    }
+
+    // the first term determines the sentence type
+    if (_term_number == 0) {
+        if (strcmp(_term, "AIVDM") == 0) {
+            // we found the sentence type for AIS
+            _sentence_valid = true;
+        }
+        return false;
+    }
+
+    // if this is not the sentence we want then wait for another
+    if (!_sentence_valid) {
+        return false;
+    }
+
+    switch (_term_number) {
+        case 1:
+            _incoming.total = strtol(_term, NULL, 10);
+            break;
+
+        case 2:
+            _incoming.num = strtol(_term, NULL, 10);
+            break;
+
+        case 3:
+            _incoming.ID = 0;
+            if (strlen(_term) > 0) {
+                _incoming.ID = strtol(_term, NULL, 10);
+             } else if (_incoming.num != 1 || _incoming.total != 1) {
+                // only allow no ID if this is a single part message
+                _sentence_valid = false;
+            }
+            break;
+
+        // case 4, chanel, either A or B, discarded
+
+        case 5:
+            if (strlen(_term) == 0) {
+                _sentence_valid = false;
+            } else {
+                strcpy(_incoming.payload,_term);
+            }
+            break;
+
+        //case 5, number of fill bits, discarded
+    }
+    return false;
+}
+
+// return the numeric value of an ascii hex character
+int16_t AP_AIS::char_to_hex(char a)
+{
+    if (a >= 'A' && a <= 'F')
+        return a - 'A' + 10;
+    else if (a >= 'a' && a <= 'f')
+        return a - 'a' + 10;
+    else
+        return a - '0';
+}
```

Comment:
We already have one of these in the codebase.

Reply:
yep

Reply:
.... so we don't want another copy....

That being said, we could probably make `AP_NMEA_Input` bare this statically at some stage.